fdbserver/MasterProxyServer.actor.cpp

```diff
@@ -398,53 +401,122 @@ ACTOR Future<Void> queueTransactionStartRequests(Reference<AsyncVar<ServerDBInfo
+						GetReadVersionReply rep;
+						rep.version = 1;
+						rep.locked = true;
+						req.reply.send(rep);
+						TraceEvent(SevWarnAlways, "ProxyGRVThresholdExceeded").suppressFor(60);
+					} else {
+						stats->addRequest(req.transactionCount);
+						// TODO: check whether this is reasonable to do in the fast path
+						for (auto tag : req.tags) {
+							(*transactionTagCounter)[tag.first] += tag.second;
+						}
 
-				if (req.debugID.present())
-					g_traceBatch.addEvent("TransactionDebug",
-					                      req.debugID.get().first(),
-					                      "MasterProxyServer.queueTransactionStartRequests.Before");
+						if (req.debugID.present())
+							g_traceBatch.addEvent("TransactionDebug",
+							                      req.debugID.get().first(),
+							                      "MasterProxyServer.queueTransactionStartRequests.Before");
 
-				if (systemQueue->empty() && defaultQueue->empty() && batchQueue->empty()) {
-					forwardPromise(GRVTimer,
-					               delayJittered(std::max(0.0, *GRVBatchTime - (now() - *lastGRVTime)),
-					                             TaskPriority::ProxyGRVTimer));
-				}
+						if (systemQueue->empty() && defaultQueue->empty() && batchQueue->empty()) {
+							forwardPromise(GRVTimer,
+							               delayJittered(std::max(0.0, *GRVBatchTime - (now() - *lastGRVTime)),
+							                             TaskPriority::ProxyGRVTimer));
+						}
 
-				++stats->txnRequestIn;
-				stats->txnStartIn += req.transactionCount;
-				if (req.priority >= TransactionPriority::IMMEDIATE) {
-					stats->txnSystemPriorityStartIn += req.transactionCount;
-					systemQueue->push_back(req);
-				} else if (req.priority >= TransactionPriority::DEFAULT) {
-					stats->txnDefaultPriorityStartIn += req.transactionCount;
-					defaultQueue->push_back(req);
+						++stats->systemAndDefaultTxnRequestIn;
+						if (req.priority >= TransactionPriority::IMMEDIATE) {
+							stats->txnSystemPriorityStartIn += req.transactionCount;
+							systemQueue->push_back(req);
+						} else if (req.priority >= TransactionPriority::DEFAULT) {
+							stats->txnDefaultPriorityStartIn += req.transactionCount;
+							defaultQueue->push_back(req);
+						}
+					}
 				} else {
-					// Return error for batch_priority GRV requests
-					int64_t proxiesCount = std::max((int)db->get().client.proxies.size(), 1);
-					if (batchRateInfo->rate <= (1.0 / proxiesCount)) {
-						req.reply.sendError(batch_transaction_throttled());
+					batchQueueLock->enter();
+					if (stats->batchTxnRequestIn.getValue() - stats->batchTxnRequestOut.getValue() -
+					        stats->batchTxnRequestErrors.getValue() >
+					    SERVER_KNOBS->START_TRANSACTION_BATCH_MAX_QUEUE_SIZE) {
+						auto& batchQueueHead = batchQueue->front();
+						batchQueueHead.reply.sendError(batch_transaction_throttled());
+						// Pop the front of batch queue as it's less promising than the incoming txn.
+						batchQueue->pop_front();
```

Comment:
I'm not sure there's any reason to prefer the newer ones to the older ones. Neither request is on the clock, and the old approach was first-come-first-serve. Doing it this proposed way also has the downside that we may not get immediate responses to our throttled batch requests, which was the goal.

Arguably, we could think about changing the existing code to immediately fail all outstanding batch requests instead of just new ones, but I haven't thought through all the implications of that.